src/java/htsjdk/samtools/BAMIndexer.java

```diff
@@ -71,14 +67,28 @@ public BAMIndexer(final File output, final SAMFileHeader fileHeader) {
      * @param fileHeader header for the corresponding bam file.
      */
     public BAMIndexer(final OutputStream output, final SAMFileHeader fileHeader) {
+        this(fileHeader, numRefs -> new BinaryBAMIndexWriter(numRefs, output));
+    }
 
+    /*
+     * Prepare to index a BAM.
+     *
+     * @param fileHeader header for the corresponding bam file.
+     * @param  createWrite a lambda that, given an Integer numReferences value, will create a BinaryBAMIndexWriter
+     *                     with that value and an appropriate output.
+      */
+    private BAMIndexer(final SAMFileHeader fileHeader, Function<Integer, BinaryBAMIndexWriter> createWriter) {
```

Comment:
I might have made this just take a `BAMIndexWriter` directly and have the callers generate it on their end.  That seems simpler, but maybe I'm missing a reason they can't do that?